src/bootstrap/src/core/builder.rs

```diff
@@ -473,9 +473,42 @@ impl StepDescription {
             return;
         }
 
-        // Handle all PathSets.
+        let orig_paths = paths.clone();
+
+        // List of `(usize, &StepDescription, Vec<PathSet>)` where `usize` is the closest index of a path
+        // compared to the given CLI paths. So we can respect to the CLI order by using this value to sort
+        // the steps.
+        let mut steps_to_run = vec![];
+
+        let mut indexed_paths: Vec<&PathBuf> = vec![];
+
         for (desc, should_run) in v.iter().zip(&should_runs) {
             let pathsets = should_run.pathset_for_paths_removing_matches(&mut paths, desc.kind);
+
+            // This value is used for sorting the step execution order.
+            // By default, `usize::MAX` is used as the index for steps to assign them the lowest priority.
+            //
+            // If we resolve the step's path from the given CLI input, this value will be updated with
+            // the step's actual index.
+            let mut closest_index = usize::MAX;
```

Comment:
The logic is a bit convoluted. What about doing something like this instead?
```rust
for (desc, should_run) in v.iter().zip(&should_runs) {
    let orig_paths = paths.clone();
    let pathsets = should_run.pathset_for_paths_removing_matches(&mut paths, desc.kind);

    // Find the lowest index of a removed path
    let index = orig_paths.iter().position(|p| !paths.contains(p)).unwrap_or(usize::MAX);
    steps_to_run.push((index, desc, pathsets));
}
```
Then we wouldn't need `indexed_paths` or the other logic (unless I'm missing something).

Reply:
This won't work. Cloning updated `paths` in each iteration can lead to use of same indexes multiple times (e.g, we may end up using `0` more than once), so the sorting will result with an incorrect order.

Reply:
Sorry, I combined too many ideas at once :facepalm: `orig_paths` here was supposed to serve for detecting removed paths, but to get the actual position, we can use another approach:

```rust
let mut steps_to_run = vec![];
let orig_paths = paths.clone();

for (desc, should_run) in v.iter().zip(&should_runs) {
    let previous_paths = paths.clone();
    let pathsets = should_run.pathset_for_paths_removing_matches(&mut paths, desc.kind);

    // Find the first path that was removed and its original index
    let index = previous_paths
        .iter()
        .find(|p| !paths.contains(p))
        .and_then(|p| orig_paths.iter().position(|p2| p2 == p))
        .unwrap_or(usize::MAX);
    steps_to_run.push((index, desc, pathsets));
}
```
If we want to remove the paths after being found (this is probably only useful to handle duplicates in `paths`?), we could use a hash map:
```rust
let mut orig_paths: HashMap<PathBuf, usize> =
  paths.iter().enumerate().map(|(i, p)| (p.clone(), i)).collect();

for (desc, should_run) in v.iter().zip(&should_runs) {
  let previous_paths = paths.clone();
  let pathsets = should_run.pathset_for_paths_removing_matches(&mut paths, desc.kind);

  // Find the first path that was removed and its original index
  let index = previous_paths
      .iter()
      .find(|p| !paths.contains(p))
      .and_then(|p| orig_paths.remove(p))
      .unwrap_or(usize::MAX);
  steps_to_run.push((index, desc, pathsets));
}
```

Reply:
Isn't the current approach more straight-forward? Specially the first examples adds  more overhead while not really helping in terms of readability.

Reply:
I guess it's a matter of style. To me it was a bit difficult to understand the interaction of the added Vecs. In my version, the discovery of removed paths is local to the body of the for loop, so it's easier for me to understand.

But as I said, this is purely a stylistic thing. If you prefer your version, feel free to r=me :)

Reply:
What about [this](https://github.com/rust-lang/rust/compare/774af29940f790204d9da551ae4f42fe899d3d56..4ad753af915840ca9de07dd2df80419b57136872) approach? Removed the outer vector of indexed paths.

Reply:
Ok, nice, without the indexed vector it's easier to understand. The hashmap solution would be too complicated in this version, since it would need to track both path usage and the index, and that gets annoying (I tried it).

So LGTM :+1: